File: src/providers/kimi/auth/headers.ts
import { hostname, release, arch, platform } from "node:os"
import { KIMI_CLI_VERSION } from "./constants.ts"
import { getDeviceId } from "./device-id.ts"

function deviceModel(): string {
  const a = arch()
  const p = platform()
  if (p === "darwin") return `macOS ${release()} ${a}`.trim()
  if (p === "win32") return `Windows ${release()} ${a}`.trim()
  return `${p} ${release()} ${a}`.trim()
}

function asciiOnly(value: string, fallback = "unknown"): string {
  // Strip non-ASCII, which would be rejected as an HTTP header value.
  const cleaned = value.replace(/[^\x20-\x7e]/g, "").trim()
  return cleaned || fallback
}

export async function commonHeaders(): Promise<Record<string, string>> {
  const deviceId = await getDeviceId()
  return {
    "X-Msh-Platform": "kimi_cli",
    "X-Msh-Version": KIMI_CLI_VERSION,
    "X-Msh-Device-Name": asciiOnly(hostname()),
    "X-Msh-Device-Model": asciiOnly(deviceModel()),
    "X-Msh-Os-Version": asciiOnly(release()),
    "X-Msh-Device-Id": deviceId,
    "User-Agent": `KimiCLI/${KIMI_CLI_VERSION}`,
  }
}

File: src/providers/kimi/auth/headers.rs
use std::collections::HashMap;

use reqwest::header::{HeaderMap, HeaderName, HeaderValue};

use super::constants::KIMI_CLI_VERSION;
use super::device_id::get_device_id;
use crate::config;

fn device_model() -> String {
    let arch = std::env::consts::ARCH;
    let os = std::env::consts::OS;
    format!("{} {}", os, arch)
}

fn ascii_only(value: &str, fallback: &str) -> String {
    let cleaned: String = value
        .chars()
        .filter(|&c| c.is_ascii() && !c.is_control())
        .collect();
    let trimmed = cleaned.trim().to_string();
    if trimmed.is_empty() {
        fallback.to_string()
    } else {
        trimmed
    }
}

pub fn common_headers() -> Result<HashMap<String, String>, anyhow::Error> {
    let device_id = get_device_id()?;
    let hostname_str = hostname::get()
        .map(|h| h.to_string_lossy().to_string())
        .unwrap_or_else(|_| "unknown".to_string());

    let mut headers = HashMap::new();
    headers.insert("X-Msh-Platform".to_string(), "kimi_cli".to_string());
    headers.insert("X-Msh-Version".to_string(), KIMI_CLI_VERSION.to_string());
    headers.insert(
        "X-Msh-Device-Name".to_string(),
        ascii_only(&hostname_str, "unknown"),
    );
    headers.insert(
        "X-Msh-Device-Model".to_string(),
        ascii_only(&device_model(), "unknown"),
    );
    headers.insert(
        "X-Msh-Os-Version".to_string(),
        ascii_only(std::env::consts::ARCH, "unknown"),
    );
    headers.insert("X-Msh-Device-Id".to_string(), device_id);
    headers.insert(
        "User-Agent".to_string(),
        config::kimi_user_agent(&format!("KimiCLI/{KIMI_CLI_VERSION}")),
    );
    Ok(headers)
}

/// Convert stringly-typed Kimi headers into a reqwest `HeaderMap`, skipping any
/// names or values that are not valid HTTP header tokens.
pub fn header_map(headers: &HashMap<String, String>) -> HeaderMap {
    let mut map = HeaderMap::new();
    for (k, v) in headers {
        if let Ok(name) = HeaderName::from_bytes(k.as_bytes())
            && let Ok(value) = HeaderValue::from_str(v)
        {
            map.insert(name, value);
        }
    }
    map
}
